### PR TITLE
[SPARK-22162] Executors and the driver should use consistent JobIDs in the RDD commit protocol.

### PR DESCRIPTION
**Pull request by rezasafi**
I have modified SparkHadoopWriter so that executors and the driver always use consistent JobIds during the hadoop commit. Before SPARK-18191, spark always used the rddId, it just incorrectly named the variable stageId. After SPARK-18191, it used the rddId in the driver, and the stageId in the executors. With this change, spark uses a closure to create configs that will be called as soon as stageId is available in DAGScheduler. This way executors and the river will consistently uses stageId.
 In addition to the existing unit tests, a test has been added to check whether executors and the driver are using the same JobId. The test failed before this change and passed after applying this fix.

### CODE REVIEW

**Comment by vanzin** [comment]
ok to test

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: you don't need the empty parents here, the return type is automatically `Unit` if the last statement is an assignment.

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: same here.

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: our convention is to put empty *braces* `{ }` not empty parens for no-op methods.

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: style nits: 
looks like an extra set of parens inside `Array`?

don't think you need `new Integer`, you can just put the raw ints

if you've got a `=>`  in your call to `map`, `filter`, etc., than use braces, not parens

so

```scala
val pairs = sc.parallelize(Array(1, 2), 1).
    map { p => (p._1 + 1, p._2 + 1) } .filter { p => p._1 > 0 }
```

and add a comment about why you are adding these extra operations before the `save` call.

**Review by squito — CHANGES_REQUESTED** [review]
can you please update the PR description to remove the original instruction lines in the template message?  Eg. "Please fill in ..."

Also better to reference a jira in the description, not a PR number.  So I'd reword to something like

I have modified SparkHadoopWriter so that executors and the driver always use consistent JobIds during the hadoop commit.   Before SPARK-18191, spark always used the rddId, it just incorrectly named the variable stageId.  After SPARK-18191, it used the rddId in the driver, and the stageId in the executors.  With this change, spark consistently uses the rddId, and the naming in the code has also been fixed.

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (rezasafi) [reply core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: The extra parenthesis is because the array has the type of pairs. I will add another pair to make it clear. I will apply your other changes.

Comment (rezasafi) [reply core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: The extra parenthesis is because I am trying to build an array of pairs. I will add another pair to make it clear. I will apply your other comments as well. Thank you very much.

Comment (squito) [reply core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: oh, of course on the pairs, sorry I missed that.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @jiangxb1987

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala]: This is used as stageId in the commit protocol, IMO the original name is better.

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala]: ditto

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: nit:
```
val pairs = sc.parallelize(Array((1, 2), (2, 3)), 2).map { p =>
  (new Integer(p._1 + 1), new Integer(p._2 + 1))
}.filter { p => p._1 > 0 }
```

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: You don't need to ensure the stageId is equal to `rddId`, just make sure it inline with the jobId is okay.

**Review by jiangxb1987 — COMMENTED** [review]
This should be a valid fix, thanks for investigating on that.

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (rezasafi) [reply core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: Thank you very much for the review. I noticed that the timeout issue that happens during the tests is actually because of my change. I didn't see the issue locally at first, since apparently I was testing on top of 2.2 branch. It seems that the committer actually look for the stageId and if there is no stage with the provided Id it will deny the commit. Since the change here proposes Rdd.id committer will always deny the commit. I will try to see if I can find a fix for this by just a small change by the end of this week. Otherwise I will close this PR to provide a more sophisticated solution for the issue.

**Comment by rezasafi** [comment]
Sorry for the delay. It seems that to be able to commit the same rdd in different stages we need to use stageId. So the jobId and other configurations in the write method of SparkHadoopWriter should be based on the stageId of the rdd and not the rddId. I have a hacky solution for this, but I am working on a better one and will update this PR ASAP.

**Comment by squito** [comment]
@rezasafi and I talked about this for a while offline.  Its a lot trickier than this seems at first.  Here's whats going on:

1. `SparkHadoopWriter` is pretending `rddId` is `stageId`, but clearly its not.  In the simple test cases they're the same, but they can be arbitrarily different.  That is not new -- it was in the old RDD commit protocol (https://github.com/apache/spark/blob/branch-2.0/core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala#L1084).  However, SPARK-18191 changed it so that on the executors, it was using the real `stageId`, which resulted in an inconsistency.

2. We really want the `stageId`.  Reza had initially made a change to just use the `rddId` consistently, since that is what was being used before.  However, this leads to an inconsistency, because the `OutputCommitCoordinator` actually expects the `stageId`.  Previously, this seemed to work because `PairRDDFunctions.saveAsNewAPIHadoopDataset` didn't go through the `OutputCommitCoordinator` (!!! can this be right?)  and `PairRDDFunctions.saveAsHadoopDataset`, which did go through the `OutputCommitCoordinator`, always used the actual `stageId` (it didn't do any setup before the `stageId` was available).  e.g here's the code in 2.0: https://github.com/apache/spark/blob/branch-2.0/core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala#L1161
The RDD alone is *not* a good unique identifier -- the same RDD can get saved to hadoop multiple times, and those could even be run in parallel.  We need to make sure those tasks don't get mixed up with each other.

3. Its actually quite hard to get the right `stageId`.  Its not exposed at all to the user before the job starts.  In fact, we *can't* expose the `stageId` before your call to `sc.runJob()`, because there could other threads racing to submit jobs, and the `stageId` isn't known until its inside the `DAGScheduler`'s eventLoop here: https://github.com/apache/spark/blob/4f8dc6b01ea787243a38678ea8199fbb0814cffc/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L866

One way to do this would be to add another version of `sc.runJob()` which let you pass in yet another function, which is run inside the `DAGScheduler` after it has the `stageId`.  I don't even think we should expose that function publicly, we just need it inside inside PairRDDFunctions.

Another possibility would be to forget about the actual stageId, and instead just get some unique id from the `OutputCommitCoordinator`, which is then passed along inside the closures.  But then you'd have to figure out how to replicate `OutputCommitCoordinator.stageStart() / stageEnd()`, which probably boils down to the same thing -- another version of `runJob`, this one which pushes the given ID down into the scheduler.

Any other ideas?

@tgravescs @markhamstra you may be interested in this as well.  I know you've told me you've had success w/ speculation lately, but seems like it would be pretty broken given the `rddId` vs. `stageId` (or maybe it doesn't matter b/c in most cases the Hadoop output committer ignores the `jobContext` passed in on `setupJob`?)

**Comment by tgravescs** [comment]
I haven't seen any issues with speculation, most people aren't on 2.2 yet though either.   Unfortunately won't have time to look at this today in detail, will try to early next week.

**Comment by squito** [comment]
also @mridulm I think you worked on the committer protocol before, both in spark and MR, so would be great to get your thoughts.

**Comment by vanzin** [comment]
High level question: do you need to create the commit id based on the rdd + stage ids, or could it be something else (like a monotonically increasing value, or a uuid)? That would probably simplify the code.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by rezasafi** [comment]
Thank you very much, @dongjoon-hyun. I forgot to pushed the fix for streaming failure. 
@vanzin My understanding based on reading the code before this change is that the logic behind it avoids multiple committing of a single stage. So I think based on that logic it should use stageId somehow.

**Comment by squito** [comment]
Jenkins, retest this please

**Comment by mridulm** [comment]
I agree with @vanzin - this looks very complicated for a enforcing a fairly simple constraint.
It would be easier to depend on a AtomicInteger in driver for the id - and propagate that to executors via Configuration (so that driver and executors use the same id).

Will that not be simpler @rezasafi  ?

**Comment by rezasafi** [comment]
Thank you very much, @mridulm for the comment. Yeah, I agree that the provided solution is complex. The complexity came form the fact that in this solution both executors and the driver use a jobId that is based on the stageId. I tried a simpler solution in October that uses rddId as a unified Id between drivers and executors, but that solution failed since committer was expecting a stageId as the jobId. Let me discuss this with more details.
In the current code there is a call to commitTask (the taskContext here is created by using stageID as the jobId):
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala#L134
That call will ended up reaching here:
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala#L172
This will follows to the following point:
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/mapred/SparkHadoopMapRedUtil.scala#L73
Which will call canCommit and canCommit will check whether the jobId is an actual stageID:
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala#L89

If we want to use a simpler solution for creating the jobId based on a monotonically increasing value then we need to make some changes in other parts of the code to remove the need for using a stageId. I need to check whether this will actually create a simpler fix (especially with regards to SPARK-4879). 
I will update here as soon as I finished investigating this. Thank you again.

**Comment by mridulm** [comment]
Internally, spark can continue to use stageId for its commit protocol - which is independent of what we expose as jobId to mapred/mapreduce committer.
As an implementation detail, we were making using of stageId in executor and rddId at driver.
Spark does not really care about what is presented to hadoop committer, and hadoop committer does not really care about what spark is using internally; as long as both are used consistently.

Will not that help make things simpler ? Or am I missing something here ?

**Comment by rezasafi** [comment]
@mridulm thank you very much for the comment. I was thinking about something similar, but more like sending both stageID and a monotonically increasing jobId to the commit method and then let try the canCommit use the stageId while we use the jobID for hadoop commits. I am testing this right now.

**Comment by rezasafi** [comment]
I am closing this PR in favor of another one with a simpler approach that I will open shortly. Thank you everyone for the comments, here.

**Comment by rezasafi** [comment]
The new PR is here:
https://github.com/apache/spark/pull/19848

**Pull request closed without merging**
